### ISSUE TITLE
Getting assertion on session_impl::enter_callback()

### ISSUE THREAD

**Issue by rameshrajagopal** [tatsuhiro-t/nghttp2#405]
Hi, 

My server has two sets of thread one is trying to get requests from client and route it. The other task is acting as a client and routing the client request to the servers. This client part i am getting assertion on 
the below lines: 
server: asio_client_session_impl.cc:515: void nghttp2::asio_http2::client::session_impl::enter_callback(): Assertion `!inside_callback_' failed.

I am not sure whether this is the problem with my implementation because it is happening every other time. It would be good , you can guide me on this. 

The implementation is below 

https://github.com/rameshrajagopal/nghttp2_asio_example/blob/master/server/proxy_slave.cpp

Thanks,
Ramesh

**Comment by tatsuhiro-t** [comment]
Your program accesses same http2 client session from 2 different threads (connect and request).  That does not work.

**Comment by rameshrajagopal** [comment]
Thanks for your response. You mean to say, I can not have http2 client session and server session on the same process ? or could you elaborate ? I basically want my process to act as "server" for the actual "clients", at the same act as "client" for the slaves or workers ? Is it possible to achieve this via this library ?

**Comment by tatsuhiro-t** [comment]
> You mean to say, I can not have http2 client session and server session on the same process ?

No.
I mean that same client::session object cannot be accessed from multiple threads at the same time.
The first thread started [here](https://github.com/rameshrajagopal/nghttp2_asio_example/blob/af1832879f88fc34b018b525f8a650e859e99fe0/server/proxy_slave.cpp#L180).
The second thread may access session object inside [reqDispatcher](https://github.com/rameshrajagopal/nghttp2_asio_example/blob/af1832879f88fc34b018b525f8a650e859e99fe0/server/proxy_slave.cpp#L196) function.  These threads run in parallel, and may access session object at the same time.

**Comment by rameshrajagopal** [comment]
I have changed my code, could you please have a look at and check the way I am doing now, is it fine ? Currently i don't see assertion failure, but previously also it was not consistent. so just wanted to check with. 
Thanks,
Ramesh

**Comment by tatsuhiro-t** [comment]
One suggestion is move L213 out of the thread lambda function:
https://github.com/rameshrajagopal/nghttp2_asio_example/blob/baff331923bb3572c3c3d013b273bc7d9c202c1d/server/proxy_slave.cpp#L213

Currently, there is a chance that push_back can be called by multiple threads at the same time, which causes undefined behaviour.

**Comment by rameshrajagopal** [comment]
I have fixed that issue by having mutex protection

**Comment by rameshrajagopal** [comment]
This is a kind of closed, i am facing another issue, I will create a separate issue for that thanks.

**Comment by VadimMolodyh** [comment]
Hi,

I have an boost::asio::io_service instance in my application that is processed in a thread pool (there are multiple threads that call io_service::run). I wanted to confirm that I'm able to reuse it when I create instance of nghttp2::asio_http2::client::session. I create session and send request under single io_service::strand and I process responses within the same callback that I receive from the session. Is it safe or do I need to have separate thread with his own io_service that is passed into session?

Thanks, Vadim
